tests/attacks/test_knockoff_nets.py

```diff
@@ -0,0 +1,434 @@
+from tests.utils_test import get_classifier_pt
+from tests.utils_test import get_iris_classifier_tf
+from tests.utils_test import get_iris_classifier_kr
+from tests.utils_test import get_iris_classifier_pt
+from art.config import ART_NUMPY_DTYPE
+
+logger = logging.getLogger(__name__)
+
+try:
+    # Conditional import of `torch` to avoid segmentation fault errors this framework generates at import
+    import torch
+    import torch.nn as nn
+    import torch.optim as optim
+except ImportError:
+    logger.info('Could not import PyTorch in utilities.')
+
+if tf.__version__[0] == '2':
+    # pylint: disable=E0401
+    import tensorflow.compat.v1 as tf
+    tf.disable_eager_execution()
+
+
+BATCH_SIZE = 100
+NB_TRAIN = 1000
+NB_EPOCHS = 10
+NB_STOLEN = 1000
+
+
+class TestKnockoffNets(unittest.TestCase):
+    """
+    A unittest class for testing the KnockoffNets attack.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        (x_train, y_train), (_, _), _, _ = load_dataset('mnist')
+
+        cls.x_train = x_train[:NB_TRAIN].astype(ART_NUMPY_DTYPE)
+        cls.y_train = y_train[:NB_TRAIN].astype(ART_NUMPY_DTYPE)
+
+    def setUp(self):
+        master_seed(1234)
+
+    def test_tfclassifier(self):
+        """
+        First test with the TensorFlowClassifier.
+        :return:
+        """
+        # Build TensorFlowClassifiers
+        victim_tfc, sess = get_classifier_tf()
+
+        # Define input and output placeholders
+        input_ph = tf.placeholder(tf.float32, shape=[None, 28, 28, 1])
+        output_ph = tf.placeholder(tf.int32, shape=[None, 10])
+
+        # Define the tensorflow graph
+        conv = tf.layers.conv2d(input_ph, 1, 7, activation=tf.nn.relu)
+        conv = tf.layers.max_pooling2d(conv, 4, 4)
+        flattened = tf.layers.flatten(conv)
+
+        # Logits layer
+        logits = tf.layers.dense(flattened, 10)
+
+        # Train operator
+        loss = tf.reduce_mean(tf.losses.softmax_cross_entropy(logits=logits, onehot_labels=output_ph))
+        optimizer = tf.train.AdamOptimizer(learning_rate=0.001)
+        train = optimizer.minimize(loss)
+
+        # TensorFlow session and initialization
+        sess.run(tf.global_variables_initializer())
+
+        # Create the classifier
+        thieved_tfc = TensorFlowClassifier(clip_values=(0, 1), input_ph=input_ph, output=logits, labels_ph=output_ph,
+                                           train=train, loss=loss, learning=None, sess=sess)
+
+        # Create random attack
+        attack = KnockoffNets(classifier=victim_tfc, batch_size_fit=BATCH_SIZE, batch_size_query=BATCH_SIZE,
+                              nb_epochs=NB_EPOCHS, nb_stolen=NB_STOLEN, sampling_strategy='random')
+        thieved_tfc = attack.extract(x=self.x_train, thieved_classifier=thieved_tfc)
+
+        victim_preds = np.argmax(victim_tfc.predict(x=self.x_train[:100]), axis=1)
+        thieved_preds = np.argmax(thieved_tfc.predict(x=self.x_train[:100]), axis=1)
```

Comment:
Would it be possible to test for the expected values of `victim_preds` and `thieved_preds`?

Reply:
For the victim_preds, it's no sense to test its expected value since it is just from the get_classifier_xx function and we made no change on it.
For the thieved_preds, it is impossible to test its expected value since we have many random factors from selecting data to the training of the thieved classifier.